File: 8searchAndRecursion/combinations.cpp
class Solution {
public:
    void combineHelper(int n, int start, int k, vector<int> &currentSet, vector<vector<int>> & collector){
        if(currentSet.size() == k){
            vector<int>* copy = new vector<int>(currentSet);
            collector.push_back(*copy);
            return;
        }
        
        if( (n - start + 1 + currentSet.size()) < k) return;

        currentSet.push_back(start);
        combineHelper(n, start+1, k, currentSet, collector);

        currentSet.pop_back();
        combineHelper(n, start+1, k, currentSet, collector);
        return;
    }

    /**
     * @param n: Given the range of numbers
     * @param k: Given the numbers of combinations
     * @return: All the combinations of k numbers out of 1..n
     */
    vector<vector<int>> combine(int n, int k) {
        vector<int> currentSet;
        vector<vector<int> collector;
        combineHelper(n, 1, k, currentSet, collector);
        return collector;
    }
};

File: 8searchAndRecursion/combinations.mjs

function combineHelper(n, k, currentSet, index, collector){
    if(currentSet.length === k) {
        collector.push([...currentSet]);
        return;
    }

    if((n.length - index + currentSet.length) < k) return;

    currentSet.push(n[index]);
    combineHelper(n, k, currentSet, index+1, collector);

    currentSet.pop();
    combineHelper(n, k, currentSet, index+1, collector);
}
/**
 * @param n: Given the range of numbers
 * @param k: Given the numbers of combinations
 * @return: All the combinations of k numbers out of 1..n
 */
export function combine (n, k) {
    let currentSet = [];
    let collector = [];
    let inputs = [];
    for(let i = 1; i <= n; i++) inputs.push(i);
    combineHelper(inputs, k, currentSet, 0, collector);
    return collector;
}


